### PR TITLE
coq_makefile2: Better error handling of failed *.v.d files

### PR DESCRIPTION
**Pull request by JasonGross**
Now we only error on failure to create *.v.d files if we are trying to build a .vo or .vio target (which are the only rules that *.v.d files add), and we don't spew warnings about not-yet-created *.v.d files.  Incidentally, this should fix the Bedrock ci.

### CODE REVIEW

**Comment by gares** [comment]
I don't understand the patch.  According to https://www.gnu.org/software/make/manual/make.html#Prerequisite-Types the meaning of order-only prerequisites is not related to failures.  It just means "the .vo depends on the .d, but if the .d is newer than the .vo don't rebuild the .vo".

I want to fix bedrock but I cant merge a commit that fixes it incidentally. What is exactly failing and why?

**Comment by JasonGross** [comment]
> the .vo depends on the .d

That's the essential part of the dependency; this means that if the target for the `.d` file fails, then the `.vo` target (and thus `make`) will also fail.

I'm looking in to figuring out why exactly bedrock is failing

**Comment by JasonGross** [comment]
Okay, here is the setup with Bedrock:
- Copying of the ml files (selected from the various compatibility options) must happen before `ocamldep` is called on, e.g., the `.mllib` files, because `ocamldep` will make the wrong `.d` file if the dependencies don't exist.  For this to work, the relevant compatibility files must exist.  The target I use to check this, called `FORCE`, also checks that all `.v` files listed in `_CoqProject.in` actually exist.  I use a single target that copies over the ml files all at once.
- Since the `.mli.d` files happen to be included after the `.v.d` files, and `make` decides to attempt to make `include`d files in reverse order, the first target that is attempted is `Bedrock/reification/extlib.mli.d`.  That tries to make `Bedrock/reification/extlib.mli`, which depends on `update-ml-files`.  The `update-ml-files` target depends on `FORCE` to ensure that all compatibility files exist (and to give a nice error message) before it tries copying one.  This fails, because `ILTac.v` doesn't exist yet.
- When the `*.mli.d` files are `-include`d, `make` will simply give up for now on making `Bedrock/reification/extlib.mli.d`.  It will eventually be asked to make ` *.v.d` file, which importantly does not depend on `FORCE` (though the `*.vo` targets do), and so making `ILTac.v` will succeed.  When it restarts the `make` (because it managed to make a `*.v.d` file), it will eventually try to make the `*.mli.d` file again, and this time, it will succeed (because now all of the `.v` files exist)
- However, when you use `include`, you never give `make` a chance to get to the `*.v.d` files, because it fails early at the `*.mli.d`

Another solution to this is to make `FORCE` depend on `ILTac.v` (and do it before the `-include`s in Bedrock's Makefile, though I'm not sure why it needs to be early...).

Regardless, I think this PR should be merged to take care of the warning spew that would otherwise be output anytime there's a missing .v.d file (e.g., after every make clean)

**Comment by JasonGross** [comment]
Wait, why remove the order-only prerequisite?  The reason to have it was to ensure that `make` would error if you were trying to build a .vo file that you can't build a .v.d file for

**Comment by gares** [comment]
Because this is not the intended use of order-only prerequisites.
The right thing is not using the `-include` but `include` instead.  Still bedrock fails, so we decided to accept the compromise. Also the verbosity of warnings was annoying, as you reported.

I think you should have a `./configure` script in bedrock that sets up things, like putting the files in place, *before* you call the makefile generate by coq_makefile.  Alternatively, if you follow the approach I recommend (not include the generated makefile but call it from another one), you could ensure things are set before calling the generated makefile.

Bottom line: the makefile generated by coq_makefile is not designed to do the job of a configure script, i.e. generate its input prerequisites (i.e. .v files) on the fly. So please, separate these two phases in the build process of your Coq developments.

**Comment by JasonGross** [comment]
> Because this is not the intended use of order-only prerequisites.

My understanding is that order-only prerequisites are to designate targets that ought to be successfully built in order to build the current target, while also saying that updating the prerequisite shouldn't force a rebuild of the current target.  It seems to me that this use-case fits that; to build `foo.vo`, you ought to have successfully built `foo.v.d`, to know what the dependencies of `foo.vo` are; at the same time, if you update `foo.v.d`, this doesn't mean you should rebuild `foo.vo`.
What's your understanding?

> Bottom line: the makefile generated by coq_makefile is not designed to do the job of a configure script, i.e. generate its input prerequisites (i.e. .v files) on the fly. So please, separate these two phases in the build process of your Coq developments.

I shall plan to do that moving forward, when I'm not trying to be compatible with legacy projects and build-systems like bedrock and fiat, or if we get around to finishing porting bedrock and fully dropping 8.4 compatibility (though it's looking more and more like Adam will declare Bedrock a finished project soon, and that bedrock-in-the-test-suite will be primarily for the code-coverage it gives the coqdevs, inasmuch as you all want that).

**Comment by maximedenes** [comment]
> (though it's looking more and more like Adam will declare Bedrock a finished project soon, and that bedrock-in-the-test-suite will be primarily for the code-coverage it gives the coqdevs, inasmuch as you all want that)

@JasonGross can you keep us informed about this? Our current policy is to test under Travis only developments that are maintained upstream.
